pkg/ruler/prom_client.go

```diff
@@ -0,0 +1,91 @@
+package ruler
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/go-kit/log/level"
+	promapi "github.com/prometheus/client_golang/api"
+	promv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql"
+
+	"github.com/cortexproject/cortex/pkg/util/spanlogger"
+)
+
+const orgIDHeader = "X-Scope-OrgId"
+
+type promClient struct {
+	promAPI promv1.API
+	timeout time.Duration
+}
+
+func NewPromClient(queryFrontendUrl, prometheusHTTPPrefix, orgID string, timeout time.Duration) *promClient {
+	if queryFrontendUrl == "" {
+		return nil
+	}
+	client, err := promapi.NewClient(promapi.Config{
+		Address:      queryFrontendUrl + prometheusHTTPPrefix,
+		RoundTripper: &addOrgIDRoundTripper{orgID: orgID, next: http.DefaultTransport},
+	})
+	if err != nil {
+		// use distributor queryable instead
+		return nil
+	}
+
+	return &promClient{
+		promAPI: promv1.NewAPI(client),
+		timeout: timeout,
+	}
+}
+
+func (p *promClient) InstantQuery(ctx context.Context, qs string, t time.Time) (promql.Vector, error) {
+	log, ctx := spanlogger.New(ctx, "promClient.InstantQuery")
+	defer log.Span.Finish()
+
+	value, warns, err := p.promAPI.Query(ctx, qs, t, promv1.WithTimeout(p.timeout))
+	if err != nil {
+		level.Error(log).Log("err", err, "query", qs)
+		return nil, err
+	}
+	if len(warns) > 0 {
+		level.Warn(log).Log("warnings", strings.Join(warns, ", "), "query", qs)
+	}
+
+	switch value.Type() {
+	case model.ValVector:
+		vector := value.(model.Vector)
+		v := make([]promql.Sample, 0, len(vector))
+		for _, sample := range value.(model.Vector) {
+			metric := make([]labels.Label, 0, len(sample.Metric))
+			for k, v := range sample.Metric {
+				metric = append(metric, labels.Label{
+					Name:  string(k),
+					Value: string(v),
+				})
+			}
+			v = append(v, promql.Sample{
+				T:      sample.Timestamp.Unix(),
+				F:      float64(sample.Value),
```

Comment:
It is possible that the returned values contain native histograms. We need to handle that.

Reply:
I going to study it, and then add handing code. Thank you.

Reply:
I don't understand "values contain native histogram".. could you give me some examples..?

Reply:
Which data are necessary for native histogram?

Reply:
Yes. But looking at the code, I don't know if there is a way to do that convertion as the returned histogram in query result only has partial information about the native histogram.

Reply:
What Thanos does it to expose a gRPC endpoint to return raw float native histograms from engine response. https://github.com/thanos-io/thanos/pull/6390

Reply:
Ah.. I got it. Should I handle it another pr or develope this pr?

Reply:
Let's see what other maintainers think before moving forward. 
We probably need a proper design document to expose that API through gRPC. If we decide to do that way, then it doesn't make sense to query query frontend any more as we need to expose that through a querier.

Reply:
ok, thank you.
@alanprot @danielblando 
Could you please take a look?